### DISCUSSION TITLE
A hotkey to insert the current date?

### DISCUSSION THREAD

**Discussion by 097115** [andmarti1424/sc-im#663] [General]
Probably, a very simple question, just can't wrap my head around it...

So, let's say I want to create a mapping that would populate the selected cell with the current date in `%Y-%m-%d` format. Where do I start?

It seems that `sc-im` alone can't solve this problem, right? So, probably calling `@ext`? or some external script? or `@lua` / lua script? would be a way to go? But how do I combine all this? :)

Really, any help will be deeply appreciated :)

**Comment by 1dancook** [comment]
Check for this entry in the help file (`:help`): `@date({date_expr}, {sexpr})`. (And the context around that entry)

Then for mapping search for `mapping` in the help. Specifically you likely want to use `:nmap`. Write your mappings in your config.

**Comment by 097115** [comment]
@1dancook, thanks for your answer!

However, `@date` still needs `{date_expr}`, right?  We have to feed it with an epoch second. And if I use `@now` as `{date_expr}` -- or some other function, like `@ext("date +'%s' #", 0)` -- it will be recalculated in the future.

The point is, I'd like to press a hotkey and automatically / programmatically get the current cell filled with the current date. Which will not change tomorrow :)

**Comment by andmarti1424** [comment]
may be adding and @if to not override if date already has been filled?
move this to discussion.

**Reply by 097115** [reply]
I'm not sure I understand you correctly?

Let's say the cell is indeed empty. Can I somehow programatically fill it with the current Epoch second? In such a way that it indeed would be a number, not a function call (so that it wouldn't change in the future)? 

As I understand, the solution provided by @1dancook still requires a user to provide the date, `{date_expr}`, manually...

**Comment by andmarti1424** [comment]
Actually what I suggested I believe it wont work. Dont really know if this is possible, even on sc.

**Reply by 097115** [reply]
Thanks for the confirmation, Andrés :) 

(So that I know that it's not just my head but that the task may actually be impossible at the moment :)

**Comment by 1dancook** [comment]
I am sorry I don't have a lot of time to play around with this myself at the moment, but I wonder if something like this might work:

Make `date.lua` in `~/.config/sc-im/lua/` with contents:
```
--get date
current_date = os.date()
--get current column and row (current cell)
current_col = sc.curcol()
current_row = sc.currow()
--write date string to current cell label
sc.lsetstr(current_col, current_row, current_date)
```

Then in sc-im you need to call the script with: `\"@lua("date.lua", 0)` which you should be able to set to a mapping.

I tried this and it doesn't work!! Theoretically we are putting a string value in the label and just getting the lua script to overwrite the label with the date string.

@andmarti1424 is there a possible bug here? 

I'm also pretty tired, maybe I missed something or there might be a way to do it similar to this.

I haven't written about this yet as a discussion topic, but I think being able to call a script as a command will be really great. `:lua scriptname.lua` -- sending all the appropriate data to the script to use (like how we can currently use `sc.curcol()` and etc. That needs a lot more thought though - but I think being able to call scripts like that independent of the cell content would be a really awesome thing.

**Comment by 1dancook** [comment]
I was able to sit down and try this again. I have it working. 

### 1- Make `date.lua` in `~/.config/sc-im/lua`

```
current_date = os.date()
current_col = sc.curcol()
current_row = sc.currow()
sc.lsetstr(current_col, current_row, current_date)
```

I don't know lua very well, but a quick online search should display some ways to manipulate the date string to your liking.

### 2-  Map something to call the script:
This is going in your scimrc file

`nmap ",d" "=@ston(@lua(\"date.lua\", 0))<CR>k=<CR>"`

Note: At the end here I am doing `k=<CR>` to effectively go back to the cell and erase the part of calling the script. At the end we should be left with just a static string in the label of the cell. If you happy to be moving to the right on `<CR>` then change it to `h=<CR>`.

**Reply by 097115** [reply]
Thanks, Daniel!

This neat trick, going back to the cell and pressing `=<CR>` to eliminate the function part, it only works for `@lua`, right? Because lau scripts _append_ the result to the cell with the formula (as a string, with `sc.lsetstr`), not just _return_ the value on enter/demand? It's really smart, and it's a pity we can't use something like with this with just `@ext`...

BTW, the original `sc` had [those macros](https://github.com/n-t-roff/sc/blob/master/SC.MACROS)... but we can't use anything like this in `sc-im` except lua, can we?

CC: @andmarti1424 :)

**Reply by 1dancook** [reply]
Yeah I am not sure the same trick will work with @ext since it will just return the value into that which you called it from. This is why as I mentioned elsewhere it'd be incredibly powerful to call scripts from command rather than just from inside a cell. If we would be able to do that we can do all kinds of fun things and the lua interface is great as well.

**Comment by 097115** [comment]
Well, there's an easier way after all :) (`h` can be replaced with `k`, of course, according to the personal settings):

    nnoremap ",d" "=@ston(@ext(\"date +'%s' #\", 0))<CR>h:valueize<CR>:format \"d%Y-%m-%d\"<CR>"

I.e., thanks to `:valueize`, we actually can eliminate the function part not only from @lua calls but any function call :)

**Reply by 1dancook** [reply]
That's a good find.
I had thought there must be a way to copy/or set just the value of the cell but couldn't remember how. I haven't used that command for anything. 
Awesome. 👌👌

**Reply by 097115** [reply]
BTW, with complex mappings that include this `h`/`k` choice, it seems a way better solution  to use `b` instead, which according to the help page, "*goes back to the previous valid cell*".

**Comment by knivey** [comment]
I found simplifying just using @now instead of externally calling date works just as well since its valueized it wont change in future.

`nnoremap ",d" "=@now<CR>h:valueize<CR>:format \"d%Y-%m-%d\"<CR>"`
